### PR TITLE
Merging configure menu with main workflow view

### PR DESCRIPTION
**Pull request by anuprulez**
This PR merges the configure workflow menu view with the main workflow view (workflow management screen). 

newconfigworkflow

When a workflow is checked in the "Show in Tools panel", it gets listed in the Tools panel and the home page refreshes.

Thanks @guerler for suggestings!

### CODE REVIEW

**Comment by mvdbeek** [comment]
The intent is awesome! Could you perhaps also remove the 'Create New Workflow' page ? Maybe the user can directly enter a workflow name after clicking on the '+'. Also, `Workflow` in the masthead does not become white anymore, it's always `Analyze Data` ... could we address that?

**Comment by mvdbeek** [comment]
@galaxybot test this

File: client/galaxy/scripts/mvc/workflow/workflow.js [hunk 1]
```diff
@@ -52,6 +53,30 @@ define( [ 'utils/utils' ], function( Utils ) {
             });
         },
 
+        // Save the workflow as an item in Tool panel
+        register_show_tool_menu: function( self ) {
```
Comment (guerler) [inline client/galaxy/scripts/mvc/workflow/workflow.js]: Parsing 'self' is not necessary here or at other locations within this module. You may just use 'this' directly within the method.

File: client/galaxy/scripts/mvc/workflow/workflow.js [hunk 2]
```diff
@@ -52,6 +53,30 @@ define( [ 'utils/utils' ], function( Utils ) {
...
+            var $el_checkboxes = self.$el.find( '.show-in-tool-panel' );
```
Comment (guerler) [inline client/galaxy/scripts/mvc/workflow/workflow.js]: Here and at similar locations you can just write `this.$( '.show-in-tool-panel' )`. `el.find` is not necessary.

File: client/galaxy/scripts/mvc/workflow/workflow.js [hunk 3]
```diff
@@ -128,8 +153,10 @@ define( [ 'utils/utils' ], function( Utils ) {
                         '<th>Owner</th>' +
                         '<th># of Steps</th>' +
                         '<th>Published</th>' +
+                        '<th>Show in Tools panel</th>' +
                     '</tr></thead>';
             _.each( workflows, function( wf ) {
+                var checkbox_html = wf.show_in_tool_panel ? '<input type="checkbox" class="show-in-tool-panel" value="'+ wf.wf_id +'" checked="'+ wf.show_in_tool_panel +'">' : '<input type="checkbox" class="show-in-tool-panel" value="'+ wf.wf_id +'"';
```
Comment (guerler) [inline client/galaxy/scripts/mvc/workflow/workflow.js]: Can you remove the redundancy in this line? I t hink it is also missing a `>` the end for the second case.

**Comment by guerler** [comment]
This is great. It makes this feature more apparent and easier to use. Thanks for implementing it.

**Comment by anuprulez** [comment]
@mvdbeek Thanks a lot for your suggestions. Can we take these changes to a separate PR?

**Comment by mvdbeek** [comment]
Sure, no problem.

File: lib/galaxy/webapps/galaxy/api/workflows.py [hunk 4]
```diff
@@ -117,33 +117,41 @@ def get_workflows_list( self, trans, for_menu, kwd ):
...
-                encoded_id = trans.security.encode_id(wf.id)
-                item['url'] = url_for('workflow', id=encoded_id)
+
+            item = wf.to_dict( value_mapper={ 'id': trans.security.encode_id } )
+            encoded_id = trans.security.encode_id(wf.id)
+            item['url'] = url_for('workflow', id=encoded_id)
             item['owner'] = wf.user.username
             item['number_of_steps'] = len( wf.latest_workflow.steps )
+            item['show_in_tool_panel'] = False
+            item['wf_id'] = wf.id
```
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/api/workflows.py]: This seems a little strange to me (but I'm not an expert, so feel free to dismiss this), you are adding both the encoded (`item['id']`) and unencoded workflow id (`item['wf_id']`).

Comment (guerler) [reply lib/galaxy/webapps/galaxy/api/workflows.py]: I agree. The unencoded id should not be returned through the api.

Comment (anuprulez) [reply lib/galaxy/webapps/galaxy/api/workflows.py]: @mvdbeek @guerler Thanks for the review!
I removed the unencoded workflow id from the api and also removed `get_workflow_menu` api call which fetched workflow ids present in the tools panel. Now, there is a flag in the workflows set itself (boolean - `show_in_tool_panel`) which determines whether any workflow is to be shown in the tools panel. Now, we just use encoded workflow id.

Comment (mvdbeek) [reply lib/galaxy/webapps/galaxy/api/workflows.py]: I don't think this is a good change, the separation between GET and PUT was fine and we should not remove things from the API. I'd say whatever consumed unencoded workflow ids should be changed to work with encoded workflow ids, which is get_workflow_menu, right?

Comment (anuprulez) [reply lib/galaxy/webapps/galaxy/api/workflows.py]: `get_workflow_menu` api is nowhere used and it used to display the unencoded workflow id. But the workflows which are present in the tools menu already have a boolean flag `show_in_tool_panel` in the list of workflows itself. But, PUT api is used to save the list using the encoded workflow ids. That's why I removed this Get API. There is already a separate api to list workflows. It was not required anymore.

Comment (mvdbeek) [reply lib/galaxy/webapps/galaxy/api/workflows.py]: How would you know this is unused ? That we don't use it in the UI doesn't mean it's not used. public APIs should not just disappear like that, without announcement or deprecation. If it is unused, you can leave a comment in the code that it is deprecated and that another endpoint should be used.

Comment (anuprulez) [reply lib/galaxy/webapps/galaxy/api/workflows.py]: I will restore the API as I am not completely sure what are all places it is used. 

But what we should do about the exposed unencoded workflow ids in the api? Shall we comment them or just remove these lines:
```
ids_in_menu = [ x.stored_workflow_id for x in user.stored_workflow_menu_entries ]
'ids_in_menu': ids_in_menu

```

Comment (mvdbeek) [reply lib/galaxy/webapps/galaxy/api/workflows.py]: So we're not actively using this endpoint ? Then I'd leave it as it is.

File: lib/galaxy/webapps/galaxy/api/workflows.py [hunk 5]
```diff
@@ -91,20 +78,25 @@ def set_workflow_menu( self, trans, **kwd ):
...
         seen_workflow_ids = set()
-        for id in workflow_ids:
-            if id in seen_workflow_ids:
-                continue
-            else:
-                seen_workflow_ids.add( id )
-            m = model.StoredWorkflowMenuEntry()
-            m.stored_workflow = q.get( id )
-            user.stored_workflow_menu_entries.append( m )
+        for wf_id in workflow_ids:
```
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/api/workflows.py]: I'd probably just decode all workflow ids using `trans.security.decode_id` as early as you can (before `sess = trans.sa_session`). The rest of the code can stay the same then.

**Comment by martenson** [comment]
@galaxybot test this

**Comment by bgruening** [comment]
@mvdbeek thanks for your review. Is this ready to go?

File: lib/galaxy/webapps/galaxy/api/workflows.py [hunk 6]
```diff
@@ -82,6 +82,10 @@ def set_workflow_menu( self, trans, **kwd ):
             workflow_ids = []
         elif type( workflow_ids ) != list:
             workflow_ids = [ workflow_ids ]
+        workflow_ids_decoded = []
```
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/api/workflows.py:85]: I'd prefer the compacter `workflow_ids = [trans.security.decode_id(wf_id) for wf_id in workflow_ids]`, but that's not a must. I do prefer not to use the plural (`ids`) in the below for loop though.

**Comment by mvdbeek** [comment]
The python part looks OK, @guerler is the rest OK for you now ?

File: client/galaxy/scripts/mvc/workflow/workflow.js [hunk 7]
```diff
@@ -32,30 +30,55 @@ define( [ 'utils/utils' ], function( Utils ) {
...
+                    var checkbox = $el_checkboxes[ item ];
+                    if( checkbox.checked ) {
+                        ids.push( checkbox.value );
+                    }
+                }
+                // Save all the checked workflows
+                $.ajax({
+                    type: 'PUT',
+                    url: Galaxy.root + 'api/workflows/menu/',
+                    data: JSON.stringify( { 'workflow_ids': ids } ),
```
Comment (guerler) [inline client/galaxy/scripts/mvc/workflow/workflow.js:70]: Is stringify necessary here?

**Comment by guerler** [comment]
It looks good. I think we can make additions with follow up PRs if necessary.

File: client/galaxy/scripts/mvc/workflow/workflow.js (hunk 7 above)
Comment (anuprulez) [reply client/galaxy/scripts/mvc/workflow/workflow.js:70]: It gives bad request without it (JSON.stringify)

**Comment by martenson** [comment]
@galaxybot test this

**Pull request merged**